classes/UpgradeTools/Module/ModuleDownloader.php

```diff
@@ -42,99 +43,86 @@ class ModuleDownloader
-
-        // Make the request
-        $context = stream_context_create([
-            'http' => [
-                'method' => 'POST',
-                'content' => 'version=' . $this->psVersion . '&method=module&id_module=' . $moduleDownloaderContext->getModuleId(),
-                'header' => 'Content-type: application/x-www-form-urlencoded',
-                'timeout' => 10,
-            ],
-        ]);
-
-        // file_get_contents can return false if https is not supported (or warning)
-        $content = Tools14::file_get_contents($addonsUrl, false, $context);
-        if (empty($content) || substr($content, 5) == '<?xml') {
-            throw (new UpgradeException($this->translator->trans('[WARNING] No response from Addons server.')))->setSeverity(UpgradeException::SEVERITY_WARNING);
+        $moduleSource = $moduleDownloaderContext->getUpdateSources()[$index];
+        $filesystem = new Filesystem();
+
+        $destinationPath = $this->downloadFolder;
+
+        if ($moduleSource->isZipped()) {
+            $destinationPath .= '/' . $moduleDownloaderContext->getModuleName() . '.zip';
+            $filesystem->copy($moduleSource->getPath(), $destinationPath);
+        } else {
+            // Module contents is already unzipped.
+            // We move it first in the sandbox folder to make sure all the files can be read.
+            $filesystem->mirror($moduleSource->getPath(), $this->downloadFolder);
         }
 
-        if (false === (bool) file_put_contents($moduleDownloaderContext->getZipFullPath(), $content)) {
-            throw (new UpgradeException($this->translator->trans('[WARNING] Unable to write module %s\'s zip file in temporary directory.', [$moduleDownloaderContext->getModuleName()])))->setSeverity(UpgradeException::SEVERITY_WARNING);
-        }
+        $this->assertDownloadedContentsIsValid($destinationPath);
+        $moduleDownloaderContext->setPathToModuleUpdate($destinationPath);
 
-        $this->logger->notice($this->translator->trans('Module %s has been successfully downloaded from Addons.', [$moduleDownloaderContext->getModuleName()]));
+        $this->logger->notice($this->translator->trans('Module %s update files (%s => %s) have been fetched from %s.', [
+            $moduleDownloaderContext->getModuleName(),
+            $moduleDownloaderContext->getReferenceVersion(),
+            $moduleSource->getNewVersion(),
+            $moduleSource->getPath(),
+        ]));
     }
 
-    private function getLocalModuleZipPath(string $name): ?string
+    /**
+     * @throws UpgradeException If download content is invalid
+     */
+    private function assertDownloadedContentsIsValid(string $destinationPath): void
     {
-        $autoUpgradeDir = _PS_ADMIN_DIR_ . DIRECTORY_SEPARATOR . 'autoupgrade';
-        $module_zip = $autoUpgradeDir . DIRECTORY_SEPARATOR . 'modules' . DIRECTORY_SEPARATOR . $name . '.zip';
+        if (is_file($destinationPath)) {
+            if (filesize($destinationPath) <= 300) {
```

Comment:
Maybe it would be interesting to store 300 in a constant, and add an explanation